## Load adversary

In [1]:
import numpy as np
from tensorflow import keras

adversary = keras.models.load_model('adv_0.52_test_r2')

2024-01-16 16:42:43.150231: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-01-16 16:42:43.150265: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-01-16 16:42:43.151138: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-01-16 16:42:43.156634: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-01-16 16:42:44.024179: W tensorflow/compiler/tf2

## Create (target) model

In [2]:
from pia_functions import train_gradient_boosting_shadow_model, data_train_test
import pandas as pd

In [3]:
# load data
X_train, X_test, y_train, y_test, sensitive, sensitive_t = data_train_test()
# create model
gb = train_gradient_boosting_shadow_model(X_train, y_train, 0)

## Generate model output

In [5]:
# load synthetic model input
model_input = pd.read_csv('data/syn_data.csv', names=X_train.columns)
# generate model output
out = gb.predict_proba(model_input)
# reformat model output
adversary_input = np.expand_dims(out.flatten(), axis=0)

In [ ]:
# next version (with new adversary):
#model_input = pd.read_csv('data/syn_data-new.csv')
#out = gb.predict_proba(model_input)
#adversary_input = np.expand_dims(out[:,0], axis=0)

## Attack

In [6]:
# generate adversary prediction
adversary.predict(adversary_input)

1/1 [==============================] - 0s 79ms/step


array([[0.73505074]], dtype=float32)

Result interpretation: The adversary has inferred that 73.5% of the target model's training data are male samples. 